moshi/src/main/java/com/squareup/moshi/ClassJsonAdapter.kt

```diff
@@ -0,0 +1,229 @@
+    } catch (e: IllegalAccessException) {
+      throw AssertionError()
+    }
+  }
+
+  override fun toJson(writer: JsonWriter, value: T?) {
+    try {
+      writer.beginObject()
+      fieldsArray.forEach { fieldBinding ->
+        writer.name(fieldBinding.name)
+        fieldBinding.write(writer, value as Any)
+      }
+      writer.endObject()
+    } catch (e: IllegalAccessException) {
+      throw AssertionError()
+    }
+  }
+
+  override fun toString() = "JsonAdapter($classFactory)"
+
+  internal open class FieldBinding<T>(val name: String, val field: Field, val adapter: JsonAdapter<T>) {
+    open fun read(reader: JsonReader, value: Any) {
+      val fieldValue = adapter.fromJson(reader)
+      field[value] = fieldValue
+    }
+
+    @Suppress("UNCHECKED_CAST") // We require that field's values are of type T.
+    open fun write(writer: JsonWriter, value: Any) {
+      val fieldValue = field[value] as T
+      adapter.toJson(writer, fieldValue)
+    }
+  }
+
+  companion object Factory : JsonAdapter.Factory {
+    override fun create(
+      type: Type,
+      annotations: Set<Annotation>,
+      moshi: Moshi
+    ): JsonAdapter<*>? {
+      if (type !is Class<*> && type !is ParameterizedType) {
+        return null
+      }
+      val rawType = type.rawType
+      if (rawType.isInterface || rawType.isEnum) return null
+      if (annotations.isNotEmpty()) return null
+      if (rawType.isPlatformType) {
+        type.throwIfIsCollectionClass(List::class.java)
+        type.throwIfIsCollectionClass(Set::class.java)
+        type.throwIfIsCollectionClass(Map::class.java)
+        type.throwIfIsCollectionClass(Collection::class.java)
+
+        var messagePrefix = "Platform $rawType"
+        if (type is ParameterizedType) {
+          messagePrefix += " in $type"
+        }
+        throw IllegalArgumentException(
+          "$messagePrefix requires explicit JsonAdapter to be registered"
+        )
+      }
+
+      require(!rawType.isAnonymousClass) {
+        "Cannot serialize anonymous class ${rawType.name}"
+      }
+      require(!rawType.isLocalClass) {
+        "Cannot serialize local class ${rawType.name}"
+      }
+      require(!(rawType.enclosingClass != null && !Modifier.isStatic(rawType.modifiers))) {
+        "Cannot serialize non-static nested class ${rawType.name}"
+      }
+      require(!Modifier.isAbstract(rawType.modifiers)) {
+        "Cannot serialize abstract class ${rawType.name}"
+      }
+      require(!rawType.isKotlin) {
+        "Cannot serialize Kotlin type " +
+          rawType.name +
+          ". Reflective serialization of Kotlin classes without using kotlin-reflect has " +
+          "undefined and unexpected behavior. Please use KotlinJsonAdapterFactory from the " +
+          "moshi-kotlin artifact or use code gen from the moshi-kotlin-codegen artifact."
+      }
+      val classFactory = ClassFactory.get<Any>(rawType)
+      val fields = sortedMapOf<String, FieldBinding<*>>()
+      var t = type
```

Comment:
As mentioned in other PRs, please use more descriptive names

Reply:
https://github.com/square/moshi/pull/1470/commits/c608415d6ab8192e49531732a66112fc8c807812